### PR TITLE
[vpp] Implement run-time HSBC

### PR DESCRIPTION
**Pull request by rkazants**
Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

File: _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp [hunk 1]
```diff
@@ -3689,6 +3689,20 @@ mfxStatus VideoVPPHW::MergeRuntimeParams(const DdiTask *pTask, MfxHwVideoProcess
             execParams->VideoSignalInfoOut.TransferMatrix = vsi->TransferMatrix;
      }
 
+    // Update ProcAmp parameters for output
+    mfxExtVPPProcAmp *procamp = reinterpret_cast<mfxExtVPPProcAmp *>(GetExtendedBuffer(outputSurf->Data.ExtParam,
```
Comment (alariono) [inline _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp]: 1. Do you have tests for it? (behavior or functional)
2. Why VPP::Reset() does not sufficient for this usage case?

Comment (rkazants) [reply _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp]: 1. I plan to do functional testing using sample_vpp
2. Using VPP::Reset is not a good way if you have a night view with a bad camera that requires per frame HSBC adjustment.

Comment (dmitryermilov) [reply _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp]: Was it a bug that HSBC changes weren't supported on fly? Or it's a new feature?

Comment (onabiull) [reply _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp]: While this doesn't use new API, this introduce undocumented change in how we use current API. Please put this code under feature define, like MFX_ENABLE_VPP_RUNTIME_HSBC and put this define in mfx_config.h under MFX_VERSION >= MFX_VERSION_NEXT guard.

File: samples/sample_vpp/src/sample_vpp.cpp [hunk 2]
```diff
@@ -769,6 +777,75 @@ int main(int argc, msdk_char *argv[])
...
+                {
+                    m_ProcAmpData[pOutSurf] = procAmp;
+                    std::vector<mfxExtBuffer*> extBuffPtrStorage;
+                    extBuffPtrStorage.push_back((mfxExtBuffer *)&m_ProcAmpData[pOutSurf]);
+                    m_extBuffPtrStorageForOutputSurf[pOutSurf] = extBuffPtrStorage;
+                    pOutSurf->Data.ExtParam = m_extBuffPtrStorageForOutputSurf[pOutSurf].data();
+                    pOutSurf->Data.NumExtParam = m_extBuffPtrStorageForOutputSurf[pOutSurf].size();
+                }
+
                 sts = frameProcessor.pmfxVPP->RunFrameVPPAsync(
```
Comment (onabiull) [inline samples/sample_vpp/src/sample_vpp.cpp]: Same comment as above. Please guard this with VERSION_NEXT.

File: _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp (hunk 1 above)
Comment (dvrogozh) [reply _studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp]: Oleg, could you, please, review and merge in this one: https://github.com/Intel-Media-SDK/MediaSDK/pull/146 before Roman will jump into changing config.h

**Comment by onabiull** [comment]
Could you please rebase? I've just merged (PR146)[https://github.com/Intel-Media-SDK/MediaSDK/pull/146], it may affect your changes. Please also either squash these commits into one, or preferably split library and samples changes into different commits.

**Comment by rkazants** [comment]
@onabiull Done with rebase and squash into the single commit.

File: samples/sample_vpp/src/sample_vpp.cpp [hunk 3]
```diff
@@ -769,6 +783,78 @@ int main(int argc, msdk_char *argv[])
...
                 }
 #endif
+
+#ifdef ENABLE_VPP_RUNTIME_HSBC
+                nOutFrames++;
+                mfxExtVPPProcAmp procAmp;
+                // set default values for ProcAmp filters
+                procAmp.Header.BufferId = MFX_EXTBUFF_VPP_PROCAMP;
+                procAmp.Header.BufferSz = sizeof(mfxExtVPPProcAmp);
+                procAmp.Brightness = 0.0F;
```
Comment (lakulako) [inline samples/sample_vpp/src/sample_vpp.cpp]: Misused 'F' literal. At least no need here.

File: samples/sample_vpp/src/sample_vpp.cpp [hunk 4]
```diff
@@ -885,6 +971,77 @@ int main(int argc, msdk_char *argv[])
             }
             else
             {
+#ifdef ENABLE_VPP_RUNTIME_HSBC
+                nOutFrames++;
```
Comment (lakulako) [inline samples/sample_vpp/src/sample_vpp.cpp]: Is it possible to avoid code duplication?

File: samples/sample_vpp/src/sample_vpp_parser.cpp [hunk 5]
```diff
@@ -199,6 +199,12 @@ msdk_printf(MSDK_STRING("   [-pa_hue  hue]        - procamp hue property.
 msdk_printf(MSDK_STRING("   [-pa_sat  saturation] - procamp satursation property. range [   0.0,  10.0] (def: 1.0)\n"));
 msdk_printf(MSDK_STRING("   [-pa_con  contrast]   - procamp contrast property.    range [   0.0,  10.0] (def: 1.0)\n"));
 msdk_printf(MSDK_STRING("   [-pa_bri  brightness] - procamp brightness property.  range [-100.0, 100.0] (def: 0.0)\n\n"));
+#ifdef ENABLE_VPP_RUNTIME_HSBC
+msdk_printf(MSDK_STRING("   [-rt_hue  interval hue1 hue2]               - procamp hue property.         range [-180.0, 180.0] (def: 0.0)\n"));
```
Comment (lakulako) [inline samples/sample_vpp/src/sample_vpp_parser.cpp]: 1. Not typical normalization.
2. Are ranges checked?

**Pull request merged**